### PR TITLE
[MXNET-938] Allow fusing weight ahead of calling forward()

### PR DESCRIPTION
**Pull request by gautham-kollu**
## Description ##
Issue : MXNET-938
Root Cause : _rnn_param_concat() takes a while to concat all the params.

This PR allows the user to call .fuse_weights() when running Gluon with ndarray there-by preprocessing the fusing-of-weights before calling forward.  Also, when running inference fusing-of-weights is not required for each batch - only once is suffice.

Some of the alternative approaches are 
 1. Fuse Parameters() when created.
 2. Fuse the weights in ".initialize()" if we are running in inference mode and if all the parameters are initialized. 

Sending a PR as a first-draft to get some early feedback.

### CODE REVIEW

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 1]
```diff
@@ -183,6 +184,16 @@ def begin_state(self, batch_size=0, func=ndarray.zeros, **kwargs):
...
 
+    def fuse_weights(self, F):
+        if F is ndarray:
+            kwargs = {i: j.data() for i, j in self._reg_params.items()}
+            params = (kwargs['{}{}_{}_{}'.format(d, l, g, t)].reshape(-1)
+                  for t in ['weight', 'bias']
+                  for l in range(self._num_layers)
+                  for d in ['l', 'r'][:self._dir]
+                  for g in ['i2h', 'h2h'])
+            self.fused_params = F.concat(*params, dim=0)
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: The fused parameter wouldn't reflect updates in the parameters.
```
In [1]: import mxnet as mx

In [2]: a = mx.nd.arange(5)

In [3]: b = mx.nd.ones((3,))

In [4]: c = mx.nd.concat(a, b, dim=0)

In [5]: c
Out[5]:

[0. 1. 2. 3. 4. 1. 1. 1.]
<NDArray 8 @cpu(0)>

In [6]: a[0] = 5

In [7]: a
Out[7]:

[5. 1. 2. 3. 4.]
<NDArray 5 @cpu(0)>

In [8]: c
Out[8]:

[0. 1. 2. 3. 4. 1. 1. 1.]
<NDArray 8 @cpu(0)>
```

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 2]
```diff
@@ -209,12 +220,16 @@ def _forward_kernel(self, F, inputs, states, **kwargs):
...
             inputs = F.swapaxes(inputs, dim1=0, dim2=1)
-        params = (kwargs['{}{}_{}_{}'.format(d, l, g, t)].reshape(-1)
-                  for t in ['weight', 'bias']
-                  for l in range(self._num_layers)
-                  for d in ['l', 'r'][:self._dir]
-                  for g in ['i2h', 'h2h'])
-        params = F._internal._rnn_param_concat(*params, dim=0)
+
+        if F is ndarray and self.fused_params is not None:
+            params = self.fused_params
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: The hybrid_forward isn't always invoked with the same context, so even if the parameter doesn't change, the context may be wrong.

**Comment by vandanavk** [comment]
@gautham-kollu Could you check the build failure and the feedback so far.

@sandeep-krishnamurthy @stu1130 for inputs/feedback

**Comment by vrakesh** [comment]
@gautham-kollu requesting to address feedback, and look into the the build failure.

**Comment by szha** [comment]
@vrakesh there's, unfortunately, no easy way to address the comment...

**Comment by Roshrini** [comment]
@gautham-kollu Based on the feedback, are you planning to come up with some other approach to fix it or close this issue?

**Comment by ankkhedia** [comment]
@gautham-kollu ping to see if there has been any update on this PR?

**Comment by kalyc** [comment]
@gautham-kollu thanks for your contribution! Could you please update the PR? If you have other priorities at the moment, feel free to close the PR and open another one when you have the time.
Thanks :)

**Pull request closed without merging**